### PR TITLE
feat(store-dir): pacquet store prune for the global virtual store (#458)

### LINKED ISSUE #458: Implement `pacquet store prune` for the global virtual store (#432 follow-up)
## Background

#432 landed Stage 1 of the global virtual store. Frozen-lockfile installs now write package contents to `<store_dir>/links/<scope>/<name>/<version>/<hash>/node_modules/<name>` and register each importer at `<store_dir>/projects/<short-hash>` → project root. Without a prune sweep, both directories grow unbounded — old slots and dead-project entries accumulate forever.

The "Out of scope" section of #432 flagged this as a dedicated follow-up.

## What needs to happen

Port upstream's [`pruneGlobalVirtualStore`](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/pruneGlobalVirtualStore.ts) and expose it via a `pacquet store prune` subcommand.

Upstream's mark-and-sweep:

1. **Get registered projects**: list `<store_dir>/projects/`, resolve each symlink target, drop entries whose target is gone. Upstream's [`getRegisteredProjects`](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/projectRegistry.ts) already has the stale-entry cleanup; pacquet only ships `register_project` from that file today (see `crates/store-dir/src/project_registry.rs`), so the read side needs porting too.
2. **Mark phase**: for each surviving project, find every `node_modules/` directory (root + workspace packages), walk every symlink that lands under `<store_dir>/links/...`, record the resolved slot path in a `reachable` set.
3. **Sweep phase**: walk `<store_dir>/links/<scope>/<name>/<version>/`, delete any `<hash>` subdirectory not in the `reachable` set. Empty `<version>/` and `<name>/` parents get cleaned up in a second pass.

## Today

- `<store_dir>/links/...` keeps growing as projects upgrade dependency versions.
- `<store_dir>/projects/<hash>` symlinks linger for projects the user has deleted.
- No CLI surface for either cleanup.

## Plan

1. Port `getRegisteredProjects` (with the stale-entry cleanup) into `pacquet_store_dir::project_registry`.
2. Add `pacquet_store_dir::prune_global_virtual_store(store_dir)` that runs the mark-and-sweep.
3. Add a `pacquet store prune` subcommand (mirror `pnpm store prune`) that calls it.
4. Tests:
   - End-to-end: register two projects, delete one project's dir on disk, run prune, assert the dead project's registry entry is gone and any slots only it referenced are gone.
   - Mark phase: workspace with two `node_modules/` dirs that both reference the same slot — slot is kept after either project is removed.
   - Sweep phase: orphan slot under `<store_dir>/links/...` with no project reference gets removed.

## Out of scope

- Pacquet's per-store-version GC (the legacy non-GVS `<store>/v11/files/` shards). That's a different cleanup with different lifetime semantics.

## References

- Upstream: [`pruneGlobalVirtualStore`](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/pruneGlobalVirtualStore.ts)
- Upstream: [`getRegisteredProjects` cleanup](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/projectRegistry.ts)
- pacquet today: `crates/store-dir/src/project_registry.rs` (the write half from #432 / #444)
- Tracked from: #432 (Out of scope)

---
Written by an agent (Claude Code, claude-opus-4-7).

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Closes #458. Ports upstream's [`pruneGlobalVirtualStore`](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/pruneGlobalVirtualStore.ts) and the read half of [`projectRegistry`](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/projectRegistry.ts#L37-L100) (pacquet shipped only the write half in #432 / #444).

The CLI wiring already existed — `StoreCommand::Prune` → `StoreDir::prune()` — but `StoreDir::prune` was a `todo!()`. With this PR `pacquet store prune` actually runs.

## What it does

**Mark phase**: walk every registered project (`<store>/projects/<short-hash>` → project root). For each project, find every `node_modules/`, follow symlinks that land under `<store>/links/...`, record reachable `<scope>/<name>/<version>/<hash>` slots, and recurse into the slot's own `node_modules/` for transitive deps.

**Sweep phase**: walk `<store>/links/<scope>/<name>/<version>/<hash>` and remove every `<hash>` not in the reachable set. Empty `<version>/`, `<name>/`, and `<scope>/` parents get cleaned up.

**Self-healing registry**: `get_registered_projects` unlinks entries whose project directory has been deleted (`ENOENT` on stat). Permission / unrelated errors surface as `PROJECT_INACCESSIBLE` / `PROJECT_REGISTRY_ENTRY_INACCESSIBLE` — matches upstream's strict stance, since silently dropping an inaccessible registry entry could remove slots the project still references.

## Notes

- Pacquet's store-dir crate doesn't have rayon plumbing yet, so the walk is sequential. Prune is one-shot (CLI), not on the hot install path; the function shape mirrors upstream's `Promise.all`-driven layout so a parallelism graft is mechanical when it lands.
- Informational messages (\"Checking N registered project(s)…\", \"Removed N package(s)…\") go to stderr via `eprintln!`. Pacquet doesn't yet thread the install-time reporter through store-dir; the proper wiring is tracked under [#344](https://github.com/pnpm/pacquet/issues/344).

## Test plan

- [x] 5 `get_registered_projects` tests: missing-registry-dir, live project, stale self-heal, mixed live/stale, dotfile-entry skip
- [x] 6 `prune` tests: no-links no-op, no-projects no-op, dead-project slot removal + empty-name-dir cleanup, shared slot survives partial unregister, orphan slot removal, transitive reachability via slot-internal `node_modules`
- [x] `just ready` (955 tests pass, 2 skipped)
- [x] `taplo format --check`
- [x] `just dylint`

---
Written by an agent (Claude Code, claude-opus-4-7).

## Summary by CodeRabbit

* **New Features**
  * Full mark-and-sweep store prune for removing orphaned package slots while preserving reachable ones.
  * Registry listing API to enumerate registered projects for store operations.
  * Cross-platform helpers to read and remove directory-shaped symlinks/junctions.

* **Bug Fixes**
  * Self-healing removal of stale registry entries and clearer error classification for inaccessible entries/targets.
  * Safer, platform-consistent handling when registering or removing symlinked registry entries.

* **Tests**
  * Expanded coverage for pruning, registry listing, mixed live/stale scenarios, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/pnpm/pacquet/pull/475)

### CODE REVIEW

File: crates/store-dir/src/prune.rs [hunk 1]
```diff
@@ -1,15 +1,503 @@
...
+    /// Behaviour:
+    /// - No `links/` directory yet: silent no-op (matches upstream's
+    ///   `if (!await pathExists(linksDir)) return`).
+    /// - No registered projects: prints `pnpm`'s informational message
+    ///   and returns. Pacquet doesn't yet thread the install-time
+    ///   reporter into store-dir, so the message goes to stderr via
+    ///   `eprintln!` until [#344] lands the proper reporter wiring.
+    /// - Otherwise: mark-and-sweep as documented in the module-level
+    ///   comment. Returns the count of removed slot directories (zero
+    ///   when nothing was unreachable).
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/prune.rs]: Doc updated in 66c0ee02. The function intentionally mirrors upstream's `Promise<void>` shape — the removed-slot count is reported to stderr to match upstream's `globalInfo("Removed N package(s) ...")` message. Wording now says exactly that.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/store-dir/src/project_registry.rs [hunk 2]
```diff
@@ -159,6 +160,186 @@ pub fn register_project(
...
+        }
+        // Only symlinks (junctions on Windows count via `is_symlink`).
+        let file_type = match entry.file_type() {
+            Ok(ft) => ft,
+            Err(_) => continue,
+        };
+        if !file_type.is_symlink() {
+            continue;
+        }
+        let link_path = entry.path();
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/project_registry.rs]: Fixed in 66c0ee02. `file_type()` errors now surface as `EntryInaccessible` with the link path attached. There's no upstream analogue — Node's `entry.isSymbolicLink()` reads a bit that's already loaded with `withFileTypes: true` and can't fail — so we err on the side of strictness rather than mirroring a non-existent error path.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/store-dir/src/prune.rs [hunk 3]
```diff
@@ -1,15 +1,503 @@
...
+/// by `walk_symlinks_to_store`'s transitive recursion instead.
+fn find_all_node_modules_dirs(project_dir: &Path) -> Vec<PathBuf> {
+    let mut out = Vec::new();
+    scan(project_dir, &mut out);
+    return out;
+
+    fn scan(dir: &Path, out: &mut Vec<PathBuf>) {
+        let Ok(entries) = fs::read_dir(dir) else {
+            return;
+        };
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/prune.rs:162]: Mixed result in 66c0ee02 — the two sites map to different upstream shapes:

- **Lines 301 / 322 (sweep phase `list_subdirs`)**: was *not* matching upstream's [`getSubdirsSafely`](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/pruneGlobalVirtualStore.ts#L282-L299), which only swallows `ENOENT` and propagates everything else. Now does the same — new `PruneError::ReadSweepDir` variant. Real bug, real fix.

- **Lines 127 / 175 (mark phase: `find_all_node_modules_dirs`, `walk_symlinks_to_store`)**: this *does* match upstream's bare `catch { return }` at [`scan`](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/pruneGlobalVirtualStore.ts#L67-L73) and [`walkSymlinksToStore`](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/pruneGlobalVirtualStore.ts#L116-L121). The concern is real (a permission error here makes prune over-aggressive), but pacquet's cardinal rule is "match pnpm's behaviour exactly" — `pacquet store prune` shares a store directory with `pnpm store prune` and the two have to agree on what counts as reachable. Diverging unilaterally would mean a slot pacquet considers unreachable could still be cleaned up by pnpm (or vice versa). Added comments at both sites flagging the deliberate match; if this should be tightened, it has to land upstream first.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/store-dir/src/project_registry.rs [hunk 4]
```diff
@@ -159,6 +160,186 @@ pub fn register_project(
...
+        } else {
+            link_path.parent().map(|p| p.join(&target)).unwrap_or_else(|| target.clone())
+        };
+
+        match fs::metadata(&absolute_target) {
+            Ok(_) => projects.push(absolute_target),
+            Err(error) if error.kind() == ErrorKind::NotFound => {
+                fs::remove_file(&link_path).map_err(|error| {
+                    GetRegisteredProjectsError::UnlinkStale { link_path: link_path.clone(), error }
+                })?;
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/project_registry.rs:331]: Fixed in 66c0ee02. Added `pacquet_fs::remove_symlink_dir` (uses `fs::remove_dir` on Windows for junctions, `fs::remove_file` on Unix for directory symlinks) and routed both this site and the heal branch in `register_project` through it. The `register_project` fix is incidental — same logical operation, same pre-existing cross-platform bug; cleaner to fix once at the helper layer than to leave the known issue in place there.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/store-dir/src/project_registry.rs [hunk 5]
```diff
@@ -159,6 +163,199 @@ pub fn register_project(
...
+            Ok(target) => target,
+            // Upstream silently skips both ENOENT and EINVAL (the
+            // "file is not a symlink" errno on Linux). EINVAL doesn't
+            // have a portable `ErrorKind` variant in stable Rust, so
+            // we match raw `errno` via `raw_os_error` when present
+            // and fall through to the generic "inaccessible" error
+            // otherwise.
+            Err(error) if is_enoent_or_einval(&error) => continue,
+            Err(error) => {
+                return Err(GetRegisteredProjectsError::EntryInaccessible { link_path, error });
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/project_registry.rs:313]: Real bug, real fix in 8c7f43c0. You're right that `std::fs::read_link` on Windows only handles `IO_REPARSE_TAG_SYMLINK` and EINVALs on junctions — every live registry entry would have been silently dropped, breaking prune entirely on Windows. ([`rust-lang/rust#28528`](https://github.com/rust-lang/rust/issues/28528) still open since 2015.)

Option **(1)** from your list — added `pacquet_fs::read_symlink_dir` that falls back to `junction::get_target` on `InvalidInput`. Routed two `fs::read_link` sites through it:
- `get_registered_projects` (the new code on this PR)
- `register_project::InspectExisting` (pre-existing bug — the heal branch was also broken on Windows since #432 / #444; fixed incidentally with the same helper)

Same pattern as `remove_symlink_dir` from the previous review pass: keep the Windows-specific syscall split contained in `pacquet_fs` instead of leaking junction handling into every call site.

Options **(2)** (switch the writer to real symlinks) and **(3)** (defer + document the limitation) would either require admin/developer-mode on Windows or leave a known-broken platform on a flagship feature. **(1)** keeps both the original `symlink_dir` semantics and the prune correctness without a privilege bump.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/store-dir/src/prune.rs [hunk 6]
```diff
@@ -1,15 +1,544 @@
...
+        return;
+    };
+    for entry in entries.flatten() {
+        let entry_path = entry.path();
+        let Ok(file_type) = entry.file_type() else {
+            continue;
+        };
+
+        if file_type.is_symlink() {
+            let Ok(target) = fs::read_link(&entry_path) else {
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/prune.rs]: Critical fix in 17d4931b. You're right — same shape as the registry-side bug from the previous review pass (#28528), but on the mark walk this time. `walk_symlinks_to_store` now goes through `pacquet_fs::read_symlink_dir`, so junction-based `node_modules/<pkg>` links are followed correctly on Windows. Before this, the mark phase would skip every direct dep link on Windows, leave `reachable` incomplete, and the sweep would delete in-use slots.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/store-dir/src/prune.rs [hunk 7]
```diff
@@ -1,15 +1,544 @@
...
+            } else {
+                entry_path.parent().map(|p| p.join(&target)).unwrap_or(target)
+            };
+            // Canonicalise both sides so a symlink-bearing path
+            // prefix doesn't fool the `starts_with` check.
+            let canonical_target =
+                dunce::canonicalize(&absolute_target).unwrap_or_else(|_| absolute_target.clone());
+            let canonical_links =
+                dunce::canonicalize(links_dir).unwrap_or_else(|_| links_dir.to_path_buf());
+            if !canonical_target.starts_with(&canonical_links) {
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/prune.rs]: Fixed in 17d4931b. Moved the `dunce::canonicalize(links_dir)` call up to `StoreDir::prune` (called once) and threaded the result down as `canonical_links` — both into the initial `walk_symlinks_to_store` call and through every recursion. Renamed the parameter so the "this is already canonical, don't re-canonicalize it" contract is visible at the signature.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/store-dir/src/project_registry.rs [hunk 8]
```diff
@@ -159,6 +167,205 @@ pub fn register_project(
...
+    },
+
+    /// Mirrors upstream's `PROJECT_REGISTRY_ENTRY_INACCESSIBLE`. Fires
+    /// only when `read_link` failed with something *other* than
+    /// `ENOENT` / `EINVAL` (those two are silently skipped to match
+    /// upstream).
+    #[display("Cannot read project registry entry {link_path:?}: {error}")]
+    #[diagnostic(
+        code(pacquet_store_dir::get_registered_projects::entry_inaccessible),
+        help("To remove this project from the registry, delete the file at: {link_path:?}")
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/project_registry.rs]: Reworded in 17d4931b. Both `EntryInaccessible` (was "delete the file at") and `ProjectInaccessible` (was "delete the symlink at") now say "delete the entry at" so the remediation guidance is accurate on Windows junctions as well as Unix symlinks.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: crates/store-dir/src/prune.rs [hunk 9]
```diff
@@ -1,15 +1,567 @@
...
+    }
+    Ok(out)
+}
+
+fn remove_dir_if_present(path: &Path) -> Result<(), PruneError> {
+    match fs::remove_dir_all(path) {
+        Ok(()) => Ok(()),
+        Err(error) if error.kind() == ErrorKind::NotFound => Ok(()),
+        Err(error) => Err(PruneError::RemoveSlot { path: path.to_path_buf(), error }),
+    }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/store-dir/src/prune.rs]: Split as suggested in fb9d110c:

- **`remove_slot_dir`** — `fs::remove_dir_all` for the actual sweep target. Subtree is known unreferenced at this point, recursive remove is correct.
- **`remove_empty_dir`** — `fs::remove_dir` with `Ok(true)` on success, `Ok(false)` for `DirectoryNotEmpty` / `NotFound` (race with a concurrent install or already gone), propagates other errors. Used for the `<version>/`, `<name>/`, `<scope>/` parents.

The counter variables also changed semantics (`removed_versions` → `emptied_versions`, `removed_pkgs` → `emptied_pkgs`): they now track *actual* removal rather than "we tried to remove", so a race that left a parent in place doesn't promote a downstream "all children gone, remove the grandparent" decision based on stale counts.

Deliberate divergence from upstream's [`rimraf`-everywhere shape](https://github.com/pnpm/pnpm/blob/94240bc046/store/controller/src/storeController/pruneGlobalVirtualStore.ts#L210-L223) — same on-disk result in the non-race case, strictly safer when the race hits. Doc on `remove_empty_dir` flags the divergence so future readers know it's intentional.

---
Written by an agent (Claude Code, claude-opus-4-7).
